#' Internal function used by readBiokitExpression
#' 
#' 
#' @param keys A vector of character strings
#' @param dfList A list of data.frames to be matched
#' @param keyColumn The column of keys in the data.frames of dfList
#' @param valueColumns One or more character strings, column names in the
#' data.frames of dfList
#' 
#' #@importFrom ribiosUtils matchColumn
#' @return A list of matrices, each containing one value column
matchKeyValue <- function(keys, dfList, keyColumn, valueColumns) {
  subdfList <- lapply(dfList, 
                      function(x) matchColumn(keys, x, keyColumn)[, valueColumns])
  resList <- lapply(valueColumns, function(col) {
    res <- do.call(cbind, lapply(subdfList, function(x) x[,col]))
    rownames(res) <- keys
    res[is.na(res)] <- 0
    return(res)
  })
  return(resList)
}



#' Read Biokit Expression files
#' 
#' 
#' @param files One or more expression files generated by the Biokit software
#' @param exprsType Which data types should become the exprs element of the
#' resulting ExpressionSet.
#' @param ngsPipelineSampleInfoFile Sample info file used by Roche
#' Bioinformatics NGS pipeline, a headerless TSV file including sample name,
#' sample group, and FASTQ files, one sample per line
#' 
#' \code{readBiokitExpression} calls read_biokit_exprs in ribiosIO to read
#' BioKit expression data, and combine data from multiple files into one
#' ExpressionSet object
#' 
#' # @importFrom ribiosIO read_biokit_exprs # @importClassesFrom Biobase
#' ExpressionSet AnnotatedDataFrame
#' @examples
#' 
#' biokitFiles <- system.file("extdata/biokit_expression_files",
#'   sprintf("biokit-output-%d.expression", 1:4),
#'   package="ribiosNGS")
#' biokitEset <- readBiokitExpression(biokitFiles)
#' 
#' @export readBiokitExpression
readBiokitExpression <- function(files,
                                 exprsType=c("ReadCount_UniqMap", "RPKM_UniqMap",
                                             "ReadCount_MultiMap", "RPKM_MultiMap",
                                             "AllMappingReads"),
                                 ngsPipelineSampleInfoFile=NULL) {
  exprsType <- match.arg(exprsType)
  tbls <- lapply(files, read_biokit_exprs)
  featTbls <- lapply(tbls, function(x) {
    annoCols <- grepl("^Annotation", colnames(x))
    res <- data.frame(Feature=rownames(x),
                      x[,annoCols])
    return(res)
  })
  featTbl <- unique(do.call(rbind, featTbls))
  uniqFeatures <- featTbl$Feature
  rownames(featTbl) <- uniqFeatures
  valueCols  <- c("RPKM_MultiMap","ReadCount_MultiMap",
                  "RPKM_UniqMap","ReadCount_UniqMap",
                  "MultiProp", "AllMappingReads")
  valueNames <- valueCols; valueNames[valueCols == exprsType] <- "exprs"
  valueMats <- matchKeyValue(uniqFeatures, tbls, 0L, valueCols)
  resEnv <- new.env()
  sapply(seq(along=valueCols), function(i) assign(x=valueNames[i], 
                                                  value=valueMats[[i]],
                                                  envir=resEnv))
  res <- new("ExpressionSet",
             assayData=resEnv,
             featureData=new("AnnotatedDataFrame", featTbl))
  sampleNames(res) <- basefilename(files)
  if(!is.null(ngsPipelineSampleInfoFile)) {
    sampleInfo <- read.table(ngsPipelineSampleInfoFile, head=FALSE,
                             col.names=c("SampleName", "SampleGroup", "Fastq1", "Fastq2"))
    sampleInfo <- matchColumn(sampleNames(res), sampleInfo, "SampleName")
    rownames(sampleInfo) <- sampleNames(res)
    pData(res) <- sampleInfo
  }
  return(res)
}


#' Read GCT files from Biokit output directory
#' 
#' 
#' @param dir Biokit output directory
#' @param anno Annotation type, either \code{refseq} or \code{ensembl} is
#' supported
#' @param type GCT file type, \code{count}, \code{rpkm}, \code{uniqCount},
#' \code{uniqRpkm}, \code{star-counts}, \code{star-rpkms}, and \code{star-tpms} are supported.
#' @return A numeric matrix with the attribute \code{desc} encoding the values
#' in the description column of the GCT format
#' @examples
#' 
#' ##... (TODO: add a mock output directory in testdata)
#' 
#' @export readBiokitGctFile
readBiokitGctFile <- function(dir, 
                              anno=c("refseq", "ensembl"),
                              type=c("counts", "rpkms", "uniqCounts", "uniqRpkms",
                                     "tpms",
                                     "star-counts",
                                     "star-rpkms",
                                     "star-tpms")) {
  anno <- match.arg(anno)  
  type <- match.arg(type)
  
  annoPath <- switch(anno,
                     "refseq"="gct",
                     "ensembl"="gct-ens")
  
  gctDir <- file.path(dir, annoPath) 
  assertDir(gctDir)
  
  filePattern <- switch(type,
                        "counts"=".*_counts.gct",
                        "rpkms"=".*_rpkms.gct",
                        "uniqCounts"=".*uniq-counts.gct",
                        "uniqRpkms"=".*uniq-rpkms.gct",
                        "tpms"=".*_tpms.gct",
                        "star-counts"=".*_star-counts.gct",
                        "star-rpkms"=".*_star-rpkms.gct",
                        "star-tpms"=".*_star-tpms.gct")
  
  gctFile <- dir(gctDir, pattern=filePattern, full.names=TRUE)
  if(length(gctFile)==0 || !file.exists(gctFile))
    stop(paste0("GCT file with the pattern'", filePattern, "' does not exist!"))
  mat <- read_gct_matrix(gctFile)
  return(mat)
}

#' Read a Biokit output directory into a DGEList object for downstream analysis
#' 
#' 
#' @param dir Biokit output directory
#' @param anno Annotation type, either \code{refseq} or \code{ensembl} is
#' supported
#' @param type count type, either \code{count} or \code{uniqCount} are
#' supported.
#' @examples
#' 
#' ##... (TODO: add a mock output directory in testdata)
#' 
#' @export readBiokitAsDGEList
readBiokitAsDGEList <- function(dir, 
                                anno=c("refseq", "ensembl"),
                                countType=c("counts", "uniqCounts", "star-counts"),
                                tpmType=c("tpms", "star-tpms")) {
  ## read gct file
  anno <- match.arg(anno)
  countType <- match.arg(countType)
  tpmType <- match.arg(tpmType)
  countMat <- readBiokitGctFile(dir, anno=anno, type=countType)
  tpmMat <- readBiokitGctFile(dir, anno=anno, type=tpmType)
  stopifnot(identical(colnames(countMat), colnames(tpmMat)))
  if(!identical(rownames(countMat), rownames(tpmMat))) {
    warning("Count matrix and TPM matrix have different row names! TPM matrix is newly organized")
    commonFeat <- intersect(rownames(countMat), rownames(tpmMat))
    tpmUniqFeat <- setdiff(rownames(countMat), rownames(tpmMat))
    tpmNewMat <- countMat
    tpmNewMat[commonFeat,] <- tpmMat[commonFeat,]
    tpmNewMat[tpmUniqFeat,] <- NA
    tpmMat <- tpmNewMat
  }

  
  ## read sample annotation, either from annot/phenoData.meta or samples.txt
  phenoDataFile <- file.path(dir, "annot", "phenoData.meta")
  samplesFile <- file.path(dir, "samples.txt")
  if(file.exists(phenoDataFile)) {
    annot <- ribiosIO::readTable(phenoDataFile, row.names=FALSE)
  } else if (file.exists(samplesFile)) {
    annot <- ribiosIO::readTable(samplesFile, row.names=FALSE)
    colnames(annot)[1:2] <- c("SampleID", "group")
    annot <- cbind(SampleName=paste(as.character(annot[, 1]),
                                as.character(annot[, 2]), sep="_"),
                   annot)
  } else {
    stop("No sample annotation was found. Contact the developer.")
  }
  
  ## to have consistent formats of sample annotation, we rename the frist three columns of annot
  colnames(annot)[1:3] <- c("SampleName", "SampleID", "group")
  annotSampleName <- as.character(annot[, 1L])
  annotSampleId <- annot[, 2L]
  annotSampleGroup <- annot[, 3L]
  if(!setequal(as.character(annotSampleName) , colnames(countMat))) {
    stop("SampleID-group and gct file sample names do not match. Contact the developer.")
  } else {
    countMat <- countMat[, annotSampleName, drop=FALSE]
    tpmMat <- tpmMat[, annotSampleName, drop=FALSE]
  }
  
  genes <- data.frame(GeneID=rownames(countMat), GeneSymbol=ribiosIO::gctDesc(countMat),
                      stringsAsFactors = FALSE)
  
  res <- DGEList(counts=countMat, samples=annot, genes=genes, group = annotSampleGroup)
  res$tpm <- tpmMat
  res$BiokitAnno <- anno
  res$BiokitCountType <- countType
  res$BiokitTpmType <- tpmType
  return(res)
}
